core/src/main/scala/pureconfig/ConfigCursor.scala

```diff
@@ -66,7 +66,33 @@ sealed trait ConfigCursor {
    *         otherwise.
    */
   def asListCursor: Either[ConfigReaderFailures, ConfigListCursor] =
-    castOrFail(LIST, Set(LIST), _.asInstanceOf[ConfigList]).right.map(ConfigListCursor(_, pathElems))
+    if (isUndefined) {
+      failed(KeyNotFound.forKeys(path, Set()))
+    } else {
+      val ve = value.valueType() match {
+        case ConfigValueType.OBJECT =>
+          val ll =
+            value.asInstanceOf[ConfigObject].asScala.foldLeft[Either[List[String], List[(Int, ConfigValue)]]](Right(Nil)) {
+              case (acc, (str, v)) =>
+                (acc, Try(str.toInt)) match {
+                  case (Right(a), Success(b)) => Right(a :+ (b -> v))
+                  case (Left(keys), Failure(_)) => Left(keys :+ str)
+                  case (_, Failure(_)) => Left(List(str))
```

Comment:
I'm inclined to follow the HOCON spec on this. I think we can be opinionated in our readers and writers but strive to follow the spec on cursors. @ruippeixotog and @leifwickland: what do you think?

Reply:
I agree. In fact, we have a lot of HOCON idiosyncrasies like this throughout our readers that should be concentrated on a lower-level layer like `ConfigCursor`. In the same way, we have some HOCON idiosyncrasies that we are not following because we try to be strict with the config value types we're given.

The HOCON spec has a whole section about [Automatic type conversions](https://github.com/lightbend/config/blob/master/HOCON.md#automatic-type-conversions), and it's implemented by Typesafe Config's [DefaultTransformer](https://github.com/lightbend/config/blob/master/config/src/main/java/com/typesafe/config/impl/DefaultTransformer.java) class. Since the class is package-private, I suggest we write our own implementation (probably using `Either` for errors) and start using it on every typed access through `ConfigCursor`. What do you think, @jcazevedo?

Reply:
I like that idea. I'll update the PR accordingly to support the type conversions `DefaultTransformer` provides in `ConfigCursor`.